components/raftstore/src/store/peer_storage.rs

```diff
@@ -767,12 +821,188 @@ where
         Ok(())
     }
 
+    pub fn clean_async_fetch_res(&mut self, low: u64) {
+        self.async_fetch_results.borrow_mut().remove(&low);
+    }
+
+    pub fn set_async_fetch_res(&mut self, low: u64, res: Option<RaftlogFetchResult>) {
+        let in_fetching = match self.async_fetch_results.borrow().get(&low) {
+            Some(RaftlogFetchState::Fetched(_)) => false,
+            Some(RaftlogFetchState::Fetching) => true,
+            None => false,
+        };
+
+        match res {
+            Some(res) => {
+                if let Some(RaftlogFetchState::Fetched(prev)) = self
+                    .async_fetch_results
+                    .borrow_mut()
+                    .insert(low, RaftlogFetchState::Fetched(res))
+                {
+                    warn!(
+                        "unconsumed async fetch res";
+                        "region_id" => self.region.get_id(),
+                        "peer_id" => self.peer_id,
+                        "res" => ?prev,
+                        "low" => low,
+                    );
+                }
+            }
+            None => {
+                if !in_fetching {
+                    let prev = self.async_fetch_results.borrow_mut().remove(&low);
+                    if prev.is_some() {
+                        self.raftlog_fetch_stats.fetch_invalid.update(|m| m + 1);
+                    }
+                }
+            }
+        }
+    }
+
+    fn async_fetch(
+        &self,
+        region_id: u64,
+        low: u64,
+        mut high: u64,
+        max_size: u64,
+        context: GetEntriesContext,
+        buf: &mut Vec<Entry>,
+    ) -> raft::Result<usize> {
+        if let Some(RaftlogFetchState::Fetching) = self.async_fetch_results.borrow().get(&low) {
+            // already an async fetch in flight
+            return Err(raft::Error::Store(
+                raft::StorageError::LogTemporarilyUnavailable,
+            ));
+        }
+
+        let tried_cnt = if let Some(RaftlogFetchState::Fetched(res)) =
+            self.async_fetch_results.borrow_mut().remove(&low)
+        {
+            assert_eq!(res.low, low);
+            let mut ents = res.ents?;
+            if ents.is_empty() {
+                return Ok(0);
+            }
+            let first = ents.first().map(|e| e.index).unwrap();
+            assert_eq!(first, res.low);
+            let last = ents.last().map(|e| e.index).unwrap();
+
+            if last >= high - 1 {
```

Comment:
I think it's more intuitive, as the index of the last entry is >= the largest index that high limit wants (high - 1)

Reply:
It's the same. `[low, high)` means `low <= last < high`, so `last + 1 <= high`. I suggest to use plus is because underflow is more likely to happen generally.